### ISSUE TITLE
too many requests error again

### ISSUE THREAD

**Issue by DChevrier1** [aquasecurity/trivy-action#430]
We upgraded to 0.28 when it came out and these errors went away, but they have come back again.  What can we do to help fix this as it is part of our CI/CD pipeline?  Thanks

ERROR    [vulndb] Failed to download artifact    repo="[ghcr.io/aquasecurity/trivy-db:2](http://ghcr.io/aquasecurity/trivy-db:2)" err="oci download error: failed to fetch the layer: GET https://ghcr.io/v2/aquasecurity/trivy-db/blobs/sha256:c3afeb28c808216cc2fa69d1e682164efd3c9967451a061778329ce94ce1a069: TOOMANYREQUESTS: retry-after: 217.944µs, allowed: 44000/minute"
[190](https://github.com/XXXX/XXXXXXXX/actions/runs/11685587597/job/32539501875#step:3:200)2024-11-05T13:55:48Z    FATAL    Fatal error    init error: DB error: failed to download vulnerability DB: OCI artifact error: failed to download vulnerability DB: failed to download artifact from any source

**Comment by ACipkowski1** [comment]
Thank you for looking into this

**Comment by JBrown413** [comment]
I'd love to see a fix!!

**Comment by CDixson1** [comment]
Received same error, TOOMANYREQUESTS. A fix would be greatly appreciated. Thanks.

**Comment by RBlanc1** [comment]
A fix for this would be helpful in our pipeline process - thx

**Comment by tliebert1** [comment]
Have run into this problem also.

**Comment by ogruene** [comment]
Hi @DChevrier1,

as far as I see, you can simply add the following env to your `aquasecurity/trivy-action` step:
```
env:
    TRIVY_DB_REPOSITORY: "public.ecr.aws/aquasecurity/trivy-db:2"
```

This should work - at least when using current `aquasecurity/trivy-action@master`. Not sure if it also works with the last release.

<sub>Oliver Grüneberg \<oliver.grueneberg@mercedes-benz.com\>, Mercedes-Benz Tech Innovation GmbH
[Provider Information](https://github.com/mercedes-benz/foss/blob/master/PROVIDER_INFORMATION.md)</sub>

**Comment by arareko** [comment]
Seems related to https://github.com/aquasecurity/trivy-action/issues/107

**Comment by ogruene** [comment]
As far as I see, it's a DB download error rather related #389

**Comment by arareko** [comment]
> As far as I see, it's a DB download error rather related #389

@ogruene Yes, both are related.

**Comment by RichardoC** [comment]
> Hi @DChevrier1,
> 
> as far as I see, you can simply add the following env to your `aquasecurity/trivy-action` step:
> 
> ```
> env:
>     TRIVY_DB_REPOSITORY: "public.ecr.aws/aquasecurity/trivy-db:2"
> ```
> 
> This should work - at least when using current `aquasecurity/trivy-action@master`. Not sure if it also works with the last release.
> 
> Oliver Grüneberg <[oliver.grueneberg@mercedes-benz.com](mailto:oliver.grueneberg@mercedes-benz.com)>, Mercedes-Benz Tech Innovation GmbH [Provider Information](https://github.com/mercedes-benz/foss/blob/master/PROVIDER_INFORMATION.md)

Mind providing a source for that repo? https://pkg.go.dev/github.com/aquasecurity/trivy-db#readme-download-the-vulnerability-database only refers to the github container registry

**Comment by brianjmurrell** [comment]
So, it seems that 0.26 was supposed to have added caching and the README says specifically: _ The action has a built-in functionality for caching and restoring [the vulnerability DB]_.

While I am seeing some caching happening in our runs:

```
Run aquasecurity/trivy-action@915b19bbe73b92a6cf82a1bc12b087c9a19a5fe2
Run aquasecurity/setup-trivy@v0.2.1
Run echo "dir=$HOME/.local/bin/trivy-bin" >> $GITHUB_OUTPUT
Run actions/cache@v4
Cache Size: ~36 MB (37525313 B)
/usr/bin/tar -xf /home/runner/work/_temp/bbc2d040-9129-4c68-941e-38abe1985a7d/cache.tzst -P -C /home/runner/work/[redacted]/[redacted] --use-compress-program unzstd
Cache restored successfully
Cache restored from key: trivy-binary-v0.56.1-Linux-X64
Run echo /home/runner/.local/bin/trivy-bin >> $GITHUB_PATH
Run echo "date=$(date +'%Y-%m-%d')" >> $GITHUB_OUTPUT
Run actions/cache@v4
Cache Size: ~38 MB (39917276 B)
/usr/bin/tar -xf /home/runner/work/_temp/aa0062e0-a74f-49e4-8884-48eb8665af21/cache.tzst -P -C /home/runner/work/[redacted]/[redacted] --use-compress-program unzstd
Received 39917276 of 39917276 (100.0%), 38.0 MBs/sec
Cache restored successfully
Cache restored from key: cache-trivy-2024-11-13
```

when it comes to getting the vulnerabilities DB, the cache does not seem to be getting used:

```
2024-11-18T14:51:53Z	INFO	[vulndb] Need to update DB
2024-11-18T14:51:53Z	INFO	[vulndb] Downloading vulnerability DB...
2024-11-18T14:51:53Z	INFO	[vulndb] Downloading artifact...	repo="ghcr.io/aquasecurity/trivy-db:2"
2024-11-18T14:51:53Z	ERROR	[vulndb] Failed to download artifact	repo="ghcr.io/aquasecurity/trivy-db:2" err="OCI repository error: 1 error occurred:\n\t* GET https://ghcr.io/v2/aquasecurity/trivy-db/manifests/2: TOOMANYREQUESTS: retry-after: 95.725µs, allowed: 44000/minute\n\n"
2024-11-18T14:51:53Z	FATAL	Fatal error
```

So why is caching happening only partially?  Is there a bug related to vulnerability DB caching?

**Comment by brianjmurrell** [comment]
> Afaik these containers get released every 6 hours, is it possible that those runs were 6 hours apart?

I don't think so.  It even happens when I "rerun failed jobs" in GitHub Actions repeatedly in quick succession when the jobs fail (i.e. minutes apart).

**Comment by grom72** [comment]
> > Trivy tool has a mechanism to verify that the database is not older than one day. If this is the case Trivy tries to download a new database. This mechanism does not take into consideration if the Trivy database is restored from the GitHub cache or not.
> 
> Clearly that mechanism is not working or at least is not being effective enough as I can have many runs in a day fail and even have two runs fail within minutes of each other.

IMHO, the database update shall be blocked when the database is restored from the GitHub cache.

**Comment by grom72** [comment]
It seems that to root cause is inside Trivy tool. It does not support the scenario that the database download fails but scanning could be executed based on cached resources.

**Comment by brianjmurrell** [comment]
That seems like a reasonable summary.
